creme/linear_model/ffm.py

```diff
@@ -0,0 +1,364 @@
+        intercept (float): Initial intercept value.
+        intercept_lr (optim.schedulers.Scheduler or float): Learning rate scheduler used for
+            updating the intercept. If a `float` is passed, then an instance of
+            `optim.schedulers.Constant` will be used. Setting this to 0 implies that the intercept
+            will be not be updated.
+        weight_initializer (optim.initializers.Initializer): Weights initialization scheme. Defaults
+            to ``optim.initializers.Zeros()``.
+        latent_initializer (optim.initializers.Initializer): Latent factors initialization scheme.
+            Defaults to ``optim.initializers.Normal(mu=.0, sigma=.1, random_state=random_state)``.
+        l1 (float): Amount of L1 regularization used to push weights towards 0.
+        l2 (float): Amount of L2 regularization used to push weights towards 0.
+        clip_gradient (float): Clips the absolute value of each gradient value.
+        random_state (int, ``numpy.random.RandomState`` instance or None): If int, ``random_state``
+            is the seed used by the random number generator; if ``RandomState`` instance,
+            ``random_state`` is the random number generator; if ``None``, the random number
+            generator is the ``RandomState`` instance used by `numpy.random`.
+
+    Attributes:
+        weights (collections.defaultdict): The current weights assigned to the features.
+        latents (collections.defaultdict): The current latent weights assigned to the features.
+
+    """
+
+    def __init__(self, n_factors, loss, optimizer, intercept, intercept_lr, weight_initializer,
+                 latent_initializer, l1, l2, clip_gradient, random_state):
+        self.n_factors = n_factors
+        self.loss = loss
+        self.optimizer = optim.SGD(0.01) if optimizer is None else optimizer
+        self.intercept = intercept
+
+        self.intercept_lr = (
+            optim.schedulers.Constant(intercept_lr)
+            if isinstance(intercept_lr, numbers.Number) else
+            intercept_lr
+        )
+
+        if weight_initializer:
+            self.weight_initializer = weight_initializer
+        else:
+            self.weight_initializer = optim.initializers.Zeros()
+
+        if latent_initializer:
+            self.latent_initializer = latent_initializer
+        else:
+            self.latent_initializer = optim.initializers.Normal(sigma=.1, random_state=random_state)
+
+        self.l1 = l1
+        self.l2 = l2
+        self.clip_gradient = clip_gradient
+
+        random_latents = functools.partial(
+            self.latent_initializer,
+            shape=self.n_factors
+        )
+
+        field_latents_dict = functools.partial(
+            collections.defaultdict, random_latents
+        )
+
+        self.weights = collections.defaultdict(self.weight_initializer)
+        self.latents = collections.defaultdict(field_latents_dict)
+
+    def _field(self, j):
```

Comment:
I've thought about it a bit and I don't think I like this approach. In this case I'm assuming that you're expecting dicts like this:

```python
x = {
    'user_Pamela': 1,
    'item_Television': 1,
}
```

I guess this approach is good because it add flexibility. I would however prefer this way:

```python
x = {
    'user': 'Pamela',
    'item': 'Television',
    'hour': 16
}
```

By the way how would you handle extra features such as `'hour'`?

Reply:
Working on it as discussed earlier, keep you in touch!

Reply:
After taking a look at it, here is what I propose:

We can define the following method:
```python
def _transform(self, x):
        return dict((f'{k}_{v}', 1) if isinstance(v, str) else (k, v) for k, v in x.items())
```
And apply it at the beginning of `fit_one` and `predict_one`, after running an experiment it's actually faster (39 sec against 44 on movielens 100k) than using `creme.processing.one_hot_encoder`. The only caveat for the end user is to be careful to set such variables to strings (e.g. when using ids).

No problem with extra features as `hour`:
```python
def _field(self, j):
    return j.split('_')[0]

_field('hour')
>>> 'hour'
```

Reply:
That's nice to hear. It's nice that you have a benchmark to measure the speed impact! As for the strings, I think it's fine. The API is now really intuitive!

Reply:
Great :)